### DISCUSSION TITLE
Blocking mitmproxy's async requests?

### DISCUSSION THREAD

**Discussion by blackthorne** [mitmproxy/mitmproxy#7028] [Q&A] [answered]
I wrote an addon that intercepts mitmproxy's Requests to add authentication headers to enable some utils using my proxy to operate in an authenticated context whilst not actually supporting or being aware of it. However, if the cookie headers become invalid, I want my addon to login on the website again to get new cookies that can be used. Ideally it would put upcoming requests on hold for a little while just until it reauthenticates to get new cookies. This got me battling with the async nature of mitmproxy where new requests keep coming in before the addon is done with authentication and consequently there requests fail. I can enumerate all that I have done to lock the execution flow but I had little success so I wanted to hear if anyone has done something like this or whether it's a dead end that would require drastic changes in mitmproxy itself.

thank you

**Comment by Prinzhorn** [comment, marked as answer]
You could always go back to synchronous hooks and block them. But I think the solution to your problem is something like an async mutex so that if your "cookie fetching logic" is still running, other requests will wait for the mutex in their respective request hooks.

I'm no Python expert, here's pseudo-esque code:

```py
import asyncio

event_lock = None

async def request():
    # If one request is currently doing the cookie dance, have all others wait
    if event_lock:
        await event_lock.wait();

    if need_fetch_cookie:
        event_lock = asyncio.Event()
        await do_the_cookie_thing()
        event_lock.set()
        event_lock = None
```

Edit: maybe more like this idk this probably has some logic flaw where multiple requests will do the cookie thing and you also need an additional `asyncio.Lock()` to mutex the cookie logic

```py
import asyncio

event_lock = asyncio.Event()
# By default the lock is open.
event_lock.set()

async def request():
    # If one request is currently doing the cookie dance, have all others wait
    await event_lock.wait();

    if need_fetch_cookie:
        # Make sure all others are blocked.
        event_lock.clear()
        await do_the_cookie_thing()
        event_lock.set()
```

**Reply by blackthorne** [reply]
Thank you, I've been experimenting this and seems to do well!

**Reply by blackthorne** [reply]
One side effect of this, is that because the async context is still present, it prevents me from using libs like Playwright (fails with error `Addon error: It looks like you are using Playwright Sync API inside the asyncio loop.`) and I can't use their async API as I would have to open the lock.

**Reply by Prinzhorn** [reply]
Please explain where you want to use Playwright

**Comment by blackthorne** [comment]
I used Playwright in the code that logins in to get my cookies. It supports a sync API but it refuses to run if the asyncio loop is detected. Not ideal but I get around it by running it in a separate process.On 23 Jul 2024, at 06:19, Alexander Prinzhorn ***@***.***> wrote:﻿
Please explain where you want to use Playwright

**Reply by Prinzhorn** [reply]
> I used Playwright in the code that logins in to get my cookies.

I don't understand the problem. The code should run inside my pseudo `await do_the_cookie_thing()` so it is both async and also has the lock. So what's the problem with using the async API?

**Reply by blackthorne** [reply]
of course you did, my bad. I implied that event_lock.set() would pause the async loop and so do_the_cookie_thing() would have to be sync code but after reading the synchronization primitives I understand the wait(), set() and clear() logic you described. I was able to make this run with your proposed logic. 

thank you!
